Source/WebApi.HypermediaExtensions/WebApi/ExtensionMethods/MvcOptionsExtensions.cs

```diff
@@ -1,13 +1,39 @@
-﻿using Microsoft.AspNetCore.Mvc;
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.Extensions.DependencyInjection;
+using WebApi.HypermediaExtensions.Hypermedia.Actions;
+using WebApi.HypermediaExtensions.JsonSchema;
 using WebApi.HypermediaExtensions.Query;
+using WebApi.HypermediaExtensions.Util;
 using WebApi.HypermediaExtensions.WebApi.AttributedRoutes;
+using WebApi.HypermediaExtensions.WebApi.Controller;
 using WebApi.HypermediaExtensions.WebApi.Formatter;
 using WebApi.HypermediaExtensions.WebApi.RouteResolver;
 
 namespace WebApi.HypermediaExtensions.WebApi.ExtensionMethods
 {
     public static class MvcOptionsExtensions
     {
+        public static IMvcBuilder AddHypermediaExtensions(this IMvcBuilder builder, IServiceCollection services, 
+            HypermediaExtensionsOptions hypermediaOptions,
+            params Assembly[] controllerAndHypermediaAssemblies)
+        {
+            //TODO: register ApplicationModel as singleton and use it everywhere to replace runtime reflection 
+            builder.AddMvcOptions(o => 
+                o.AddHypermediaExtensions(controllerAndHypermediaAssemblies: controllerAndHypermediaAssemblies)
+                .AddHypermediaParameterBinders(!hypermediaOptions.ImplicitHypermediaActionParameterBinders)
+            );
+            if (hypermediaOptions.AutoDeliverNJsonSchemaForActionParameterTypes)
+                services.AutoDeliverActionParameterSchemas(controllerAndHypermediaAssemblies);
+
+            services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
```

Comment:
Should we use: TryAddSingleton? In case someon already registered?

Reply:
https://stackoverflow.com/questions/48185894/when-to-use-tryaddsingleton-or-addsingleton

Reply:
But what if somebody already registered another type? It wouldn't work then. So perhaps leave it like it is and by the exception make him remove the duplicate registration or point to the conflict?